tests/test_cases/simulation/test_mpi_ensemble.cu

```diff
@@ -0,0 +1,289 @@
+    void SetUp() override {
+        int flag = 0;
+        MPI_Finalized(&flag);
+        if (flag) {
+            // Multi node MPI run
+            GTEST_SKIP() << "Skipping single-node MPI test, MPI_Finalize() has been called";
+            return;
+        }
+        // Init MPI, fetch rank and size
+        int thread_provided = 0;
+        // MPI can only be init once, for certain test cases we do some initial MPI comms for setup
+        MPI_Initialized(&flag);
+        if (!flag) {
+            // MPI single means that only the main thread will perform MPI actions
+            MPI_Init_thread(NULL, NULL, MPI_THREAD_SINGLE, &thread_provided);
+            if (thread_provided != MPI_THREAD_SINGLE) {
+                THROW exception::UnknownInternalError("MPI unable to provide MPI_THREAD_SINGLE support");
+            }
+        }
+        MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
+        MPI_Comm_size(MPI_COMM_WORLD, &world_size);
+        // Create a shared memory split
+        MPI_Comm group;
+        MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_SHARED, 0, MPI_INFO_NULL, &group);
+        MPI_Comm_size(group, &group_size);
+        if (world_size < 1) {
+            GTEST_SKIP() << "world_size<1, something went wrong.";
+        }
+        initModel();
+    }
+    void TearDown() override {
+        if (ensemble) delete ensemble;
+        if (exit_log_cfg) delete exit_log_cfg;
+        if (plans) delete plans;
+        if (model) delete model;
+    }
+    void initModel() {
+        model = new flamegpu::ModelDescription("MPITest");
+        model->Environment().newProperty<int>("counter", -1);
+        model->Environment().newProperty<int>("init_counter", -1);
+        model->newAgent("agent");
+        model->newLayer().addHostFunction(model_step);
+    }
+    void initPlans() {
+        // 100 runs, each 10 steps which take >100ms each,
+        // therefore each sim will take >1 second.
+        // Estimate 30 seconds for full test across 4 threads
+        plans = new flamegpu::RunPlanVector(*model, 100);
+        plans->setSteps(10);
+        plans->setPropertyLerpRange<int>("counter", 0, 99);
+        plans->setPropertyLerpRange<int>("init_counter", 0, 99);
+    }
+    void initExitLoggingConfig() {
+        exit_log_cfg = new flamegpu::LoggingConfig(*model);
+        exit_log_cfg->logEnvironment("counter");
+    }
+    void initEnsemble() {
+        initPlans();
+        initExitLoggingConfig();
+        ensemble = new flamegpu::CUDAEnsemble (*model);
+        ensemble->Config().concurrent_runs = 1;
+        if (group_size == world_size) {
+            // Single node MPI run
+            int deviceCount = 0;
+            gpuErrchk(cudaGetDeviceCount(&deviceCount));
+            if (deviceCount < world_size) {
+                MPI_Finalize();
+                GTEST_SKIP() << "Skipping single-node MPI test, world size (" << world_size << ") exceeds GPU count (" << deviceCount << "), this would cause test to stall.";
```

Comment:
I think the stall was a hypothesis tbh, as init/sync things wouldn't match up.

Reply:
Yes it does stall.

The code directly above this basically for purposes of a single node test, split's the GPUs between the MPI workers. But if you have more workers than GPUs. One of more workers will be left without a GPU and hence can't start, so the first MPI sync hit will never be reached by those workers causing the stall.

As we recommend users only use 1 MPI worker per node (or simply every worker has unique access to it's GPUs), this seems acceptable. It's merely a good way to allow us to perform local testing without a complex setup to divide GPUs.

Reply:
Still concerned about the stall potentially being triggered by users in real code, but the tests skipping in this case is OK, although it would be nice to be able to test the mpi calls in a single GPU machine, even if that is a likely-to-fail scenario for sufficiently large models/ensembles.

Reply:
Whilst what you suggest would be technically possible to make possible via the test suite. We're already constrained to running these tests on Linux, which we have two multi-gpu boxes regularly sitting idle. Hence, I'm not currently bothered enough to address it. (We already saw that MPI on a single node is not a good equivalent to multi-node [e.g. the problems we had on Bede], so testing should really be performed on any system that MPI will be used).

Reply:
There's no fundamental differentce between mpi comms within and between machines, just the latency will be higher (and not using MPI might not be the most performant implementation). 

The bede mpi issues are unrelated to the MPI calls themselves, and more with the mpi version/install config (afiak). I need to dig back through bede related comms to see if they were similar issues encountered by others (I know there were MPI issues that another RSE was investigating, but need to look into it).

`mpirun -n 2` in a single GPU machine can be its own issue to go on the backlog, but I still want to get my head around the potential deadlock prior to approval.

Reply:
> mpirun -n 2 in a single GPU machine can be its own issue to go on the backlog, but I still want to get my head around the potential deadlock prior to approval.

If we do want a python test, unelss there's an mpi package the world size/rank can't be probed so any test would need to support that.

Reply:
Opened #1146. 

This scenario will need thoroguhly documenting though if it could be triggered by users.

Reply:
The stall in the test suite only occurs because it refuses to allocate runners to an ensemble if they exceed gpu count. Hence the syncs within the ensemble would never be hit.

A user not messing about with mpi ranks etc, will not hit a stall.